plugin-server/src/cdp/consumers/cdp-aggregation-writer.consumer.ts

```diff
@@ -0,0 +1,261 @@
+            const key = `${event.teamId}:${event.personId}:${event.filterHash}:${event.date}`
+            const existing = aggregatedEventsMap.get(key)
+
+            if (existing) {
+                existing.counter += 1
+            } else {
+                aggregatedEventsMap.set(key, {
+                    ...event,
+                    counter: 1,
+                })
+            }
+        }
+
+        return Array.from(aggregatedEventsMap.values())
+    }
+
+    // Process batch by aggregating and writing to postgres
+    private async processBatch(parsedBatch: ParsedBatch): Promise<void> {
+        // Deduplicate person performed events
+        const deduplicatedPersonEvents = this.deduplicatePersonPerformedEvents(parsedBatch.personPerformedEvents)
+
+        // Aggregate behavioural filter matched events
+        const aggregatedBehaviouralEvents = this.aggregateBehaviouralFilterMatchedEvents(
+            parsedBatch.behaviouralFilterMatchedEvents
+        )
+
+        // This will write both arrays in one transaction
+        await this.writeToPostgres(deduplicatedPersonEvents, aggregatedBehaviouralEvents)
+    }
+
+    // Helper to build person events CTE
+    private buildPersonEventsCTE(personEvents: PersonEventPayload[]): string {
+        const values = personEvents
+            .map((event) => `(${event.teamId}, '${event.personId}', '${event.eventName.replace(/'/g, "''")}')`)
+            .join(',')
+
+        return `person_inserts AS (
+            INSERT INTO person_performed_events (team_id, person_id, event_name)
+            VALUES ${values}
+            ON CONFLICT (team_id, person_id, event_name) DO NOTHING
+            RETURNING 1
+        )`
+    }
+
+    // Helper to build behavioural events CTE
+    private buildBehaviouralEventsCTE(behaviouralEvents: AggregatedBehaviouralEvent[]): string {
+        const values = behaviouralEvents
+            .map(
+                (event) =>
+                    `(${event.teamId}, '${event.personId}', '${event.filterHash}', '${event.date}', ${event.counter})`
+            )
+            .join(',')
+
+        return `behavioural_inserts AS (
+            INSERT INTO behavioural_filter_matched_events (team_id, person_id, filter_hash, date, counter)
+            VALUES ${values}
+            ON CONFLICT (team_id, person_id, filter_hash, date) 
+            DO UPDATE SET counter = behavioural_filter_matched_events.counter + EXCLUDED.counter
+            RETURNING 1
+        )`
+    }
+
+    // Write both event types to postgres in a single query within a transaction
+    private async writeToPostgres(
+        personEvents: PersonEventPayload[],
+        behaviouralEvents: AggregatedBehaviouralEvent[]
+    ): Promise<void> {
+        if (personEvents.length === 0 && behaviouralEvents.length === 0) {
+            return
+        }
+
+        try {
+            await this.hub.postgres.transaction(PostgresUse.COUNTERS_RW, 'aggregation-writer', async (client) => {
```

Comment:
We talked about this 😅 We don't want a 3 phase transaction. Just one simple query with the transaction built into it. By doing it this way you're needlessly doing 3 trips to PG